### DISCUSSION TITLE
How to get a distribution of constraint parameters from 'emcee_sol'?

### DISCUSSION THREAD

**Discussion by UltraX3D** [phoebe-project/phoebe2#695] [Questions]
I set three constraints  **'b.flip_constraint('pot', solve_for='requiv@primary')' , 'b.flip_constraint('fillout_factor', solve_for='pot')', 'b.flip_constraint('teffratio', solve_for='teff@secondary')'**. Then, I ran the MCMC, and I obtained the distribution of some parameters, **'q, fillout_factor, incl@binary, teffratio'** , through some codes, such as **'Q=b.get_parameter('q').get_distribution('emcee_sol')'**.

But I cannot get the distribution of constraint parameters, such as **'teff@secondary, requiv@secondary'**, so may I ask what I should do to get the distribution of constraint parameters?

**Comment by kecnry** [comment]
That should be supported, although you might need to specify `context='component'` to avoid retrieving the constraint parameter itself.  If it's still not working, can you provide a minimal script showing what you have (perhaps replacing the emcee call with attaching manual distributions)?

Alternatively, if you want a corner plot of your distributions, you can pass a list of twigs, including constrained parameters, to [b.plot_distribution_collection](http://phoebe-project.org/docs/latest/api/phoebe.frontend.bundle.Bundle.plot_distribution_collection).

**Comment by UltraX3D** [comment]
Actually，what I want to do is propagating the primary component luminosity distribution to the secondary, then derived the ratio of luminosity. But it shown that 
![error](https://user-images.githubusercontent.com/124908444/218410922-a2fc23dc-2952-4acd-89b8-7fb1fdbe1d99.jpg)

So I would like to ask you two questions.
(1) It seems to indicate that we cannot change the value of the constrained parameter, so what shouid I do to get the median and uncertainties  of secondary luminosity distribution?

(2) The arguments **'lnprob_cutoff'** does not seem to affect which part of the distribution is added, not only in the corner plot, but also in the new distribution set 'mydist1', so how do I specify which part of the distribution is added?


Here is my scipt, which is reference to this tutorial  [Propagating Distributions](http://phoebe-project.org/docs/latest/examples/distribution_pblum) .

```
bc = b.copy()
def sample_pblum_secondary(pblum_primary, teff_secondary, requiv_secondary):
    bc.set_value(qualifier='pblum', component='primary', context='dataset', value=pblum_primary)
    bc.set_value(qualifier='teff', component='secondary', context='component', value=teff_secondary)
    bc.set_value(qualifier='requiv', component='secondary', context='component', value=requiv_secondary)
    return bc.compute_pblums().get('pblum@secondary@lcT').value

lp1=bc.get_parameter('pblum@primary@lcT@lc@dataset',lnprob_cutoff=0).get_distribution('emcee_sol')
lp2=bc.get_parameter('teff@secondary@star@component',lnprob_cutoff=0).get_distribution('emcee_sol')
lp3=bc.get_parameter('requiv@secondary@star@component',context='component',lnprob_cutoff=0).get_distribution('emcee_sol')

bc.add_distribution({'pblum@primary': lp1,
                    'teff@secondary': lp2,
                    'requiv@secondary': lp3},
                   distribution='mydist1',include_constrained=True)

_ = bc.plot_distribution_collection('mydist1', show=True, lnprob_cutoff=0, 
                                    parameters=['pblum@primary','teff@secondary','requiv@secondary'], include_constrained=True)

dc, _ = bc.get_distribution_collection('mydist1', include_constrained=True)
dc.dists

pblum_sec_dist = distl.function(sample_pblum_secondary,
                                args=(lp1, lp2, lp3),
                                vectorized=False,
                                label='pblum@secondary@lcT',
                                label_latex='$L_\\mathrm{ pb, secondary, lc01 }$')

dc = distl.DistributionCollection(dc.dists[0], dc.dists[1], dc.dists[2], pblum_sec_dist)

_ = dc.plot_sample(size= 500, plot_uncertainties=False)
```


Could you give me some code modification suggestions or a workable script to get the distribution of secondary luminosity and the median,  uncertainties corresponding to the distribution.

**Reply by kecnry** [reply]
> It seems to indicate that we cannot change the value of the constrained parameter, so what shouid I do to get the median and uncertainties of secondary luminosity distribution?

Can you attach a minimal script (excluding your data - but just setting any flipping of constraints and creating distributions from scratch) to reproduce from scratch?

> The arguments 'lnprob_cutoff' does not seem to affect which part of the distribution is added, not only in the corner plot, but also in the new distribution set 'mydist1', so how do I specify which part of the distribution is added?

Yes, `get_parameter` does not accept `lnprob_cutoff`.  `lnprob_cutoff` is only applicable when processing the samples from an emcee solution, to _create_ the distributions.

**Reply by UltraX3D** [reply]
> Can you attach a minimal script (excluding your data - but just setting any flipping of constraints and creating distributions from scratch) to reproduce from scratch?

Sure, here is my script:
[example.zip](https://github.com/phoebe-project/phoebe2/files/10775568/example.zip)

> Yes, get_parameter does not accept lnprob_cutoff. lnprob_cutoff is only applicable when processing the samples from an emcee solution, to create the distributions.

Generally, the **'burn-in'** part must be deleted when calculating the physical parameters, so in my opinion, this distributions of **'pblum@primary'**,  **'teff@secondary'**, and **'requiv@secondary'** need to be treated like this before they are added to the new distribution **'mydist1'**. After we do that, then we can use  **'mydist1'** to get the secondary luminosity distribution, right?

**Question**:
Could you give me some code to delete the **'burn-in'** part, before they are added to the new distribution, and how do we get the median and uncertainties of secondary luminosity distribution from secondary luminosity distribution?

**Reply by kecnry** [reply]
Unfortunately I cannot reproduce from that script (there are undefined variables), but you should be able to pass burnin options when plotting or adopting the solution from emcee (see the bottom of the [emcee tutorial](http://phoebe-project.org/docs/latest/tutorials/emcee) for details).  The very end of that same tutorial also shows how to use [uncertainties_from_distribution_collection](http://phoebe-project.org/docs/latest/api/phoebe.frontend.bundle.Bundle.uncertainties_from_distribution_collection) to access the central value and percentile values.
